### PR TITLE
feat(docs): Add a deep dive on functional testing

### PR DESCRIPTION
**Pull request by shane-tomlinson**
@mozilla/fxa-devs - r?

### CODE REVIEW

File: docs/fxa-engineering/functional-testing.md [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+            selectors.PRIVACY_POLICY.LINK_BACK,
+            selectors.SIGNUP_PASSWORD.HEADER
+          )
+        );
+    },
+  },
+});
+```
+
+Very few low level [leadfood commands](https://theintern.io/docs.html#Leadfoot/2/api/Command) are used directly, instead we prefer to use [higher level helpers](https://github.com/mozilla/fxa/blob/master/packages/fxa-content-server/tests/functional/lib/helpers.js). Leadfoot commands are often *too* low level and end up requiring a lot
```
Comment (clouserw) [inline docs/fxa-engineering/functional-testing.md]: *leadfoot

File: docs/fxa-engineering/functional-testing.md [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+in mind that testing locally does not incur network latency. Finally, tests run on
+CircleCI and TeamCity are run in virtual machines instead of bare hardware, backend
+requests and even the test runner are often an order of magnitude slower than
+local machines.
+
+One of the most common problems is clicking on a submit button and then immediately
+checking text, an input element value, or an attribute value, of an element on a
+subsequent screen without actually ensuring the screen is visible. The high level
+helper functions *try* to take this into account, but sometimes fail. The simplest
+approach to this is to **wait for some expected DOM mutation to occurr**
```
Comment (clouserw) [inline docs/fxa-engineering/functional-testing.md]: *occur

File: docs/fxa-engineering/functional-testing.md [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+    return this.remote.then(clearBrowserState({ force: true }));
+  },
+...
+```
+
+### An element is not visible or is fading in
+
+One common problem is that Selenium *sometimes* refuses to read attribute values
+on DOM elements unless they are 100% visible, meaning elements that are in the process
+of being faded in, out sometimes cause errors. This is particularly problematic
```
Comment (clouserw) [inline docs/fxa-engineering/functional-testing.md]: I think missing an "or" in there

File: docs/fxa-engineering/functional-testing.md [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+
+The full set of functional tests is run on [CircleCI on every checkin](./fxa-tests-circleci) and on [TeamCity](https://tc-test.dev.lcip.org/) every time a pull request is merged to master.
+
+[The Intern](https://theintern.io/) library is used to run the tests, which itself is a wrapper around the [Leadfoot](https://theintern.io/docs.html#Leadfoot/2/api/Command) WebDriver library.
+
+The functional tests can seem impenetrable, don't worry, they'll become second nature after a while. They continually save our bacon, and every new feature should have corresponding functional tests.
+
+## Why are there so many test suites that seem to do the same thing?
+A close look at the tests makes it look like there are a lot of tests that do
+the same thing, and you are right. Sometimes this is intention, sometimes it's not.
```
Comment (shane-tomlinson) [inline docs/fxa-engineering/functional-testing.md]: `intention`=>`intentional`

File: docs/fxa-engineering/functional-testing.md [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+const FunctionalHelpers = require('./lib/helpers');
+const selectors = require('./lib/selectors');
+
+const ENTER_EMAIL_URL = intern._config.fxaContentRoot;
+const PP_URL = intern._config.fxaContentRoot + 'legal/privacy';
+
+const {
+  // name of helpers to import. There are helpers
+  clearBrowserState,
+  openPage,
```
Comment (shane-tomlinson) [inline docs/fxa-engineering/functional-testing.md]: Add `createEmail`

File: docs/fxa-engineering/functional-testing.md [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+)
+.then(getWebChannelMessageData('fxaccounts:login'))
+.then(messageData => {
+  assert.equal(messageData.email, 'testuser@testuser.com');
+})
+...
+```
+
+### Add a new helper
+It's easy! When adding new forms that must be completed many times throughout the test,
```
Comment (shane-tomlinson) [inline docs/fxa-engineering/functional-testing.md]: `many times throughout the test` => `in multiple tests`

File: docs/fxa-engineering/functional-testing.md [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+Responses to these can both be overridden by providing a `webChannelResponses` object
+in the `options` parameter of `openPage`. An example from [sync_v3_sign_in.js](https://github.com/mozilla/fxa/blob/abfb9822760f9edf5175bdded0efa8e074b8ed84/packages/fxa-content-server/tests/functional/sync_v3_sign_in.js#L55-L62):
+
+```js
+...
+.then(
+  openPage(ENTER_EMAIL_URL, selectors.ENTER_EMAIL.HEADER, {
+    query,
+    webChannelResponses: {
+      // simulate the user declining whethere two Sync accounts
```
Comment (shane-tomlinson) [inline docs/fxa-engineering/functional-testing.md]: `whethere`  => `whether`

**Pull request merged**